### PR TITLE
feat: add module-resolver utils

### PR DESCRIPTION
**Pull request by ashoktamang**
add class that rewrites the imports of a given file and its dependent files based on where
the file has been moved inside the project.

### CODE REVIEW

File: addon/ng2/utilities/module-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+import * as path from 'path';
+import * as ts from 'typescript';
+import * as dependentFilesUtils from './get-dependent-files';
+
+import { Promise } from 'es6-promise';
+import { Change, ReplaceChange } from './change';
+
+/** @const The root directory of Angular Project. */
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: You don't need `@const`, `@class`, `@member,`, etc.

File: addon/ng2/utilities/module-resolver.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+const ROOT_PATH = path.resolve('src/app');
+
+/** 
+ * @class Rewrites import module of dependent files when the file is moved. 
+ *        Also rewrites export module of related index file of the given file.
+ */
+export class ModuleResolver {
+
+  /** @member Description of the module resolution. */
+  description: string;
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: What is `description` used for?

File: addon/ng2/utilities/module-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+   * then apply() method is called sequentially.
+   * 
+   * @param changes {Change []} 
+   * @return Promise after all apply() method of Change class is called 
+   *         to all Change instances sequentially.
+   */
+  applySortedChangePromise(changes: Change[]): Promise<void> {
+    return changes
+      .sort((currentChange, nextChange) => nextChange.order - currentChange.order)
+      .reduce((newChange, change) => newChange.then(() => change.apply()), Promise.resolve());
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Nice!

File: addon/ng2/utilities/module-resolver.ts (hunk 2 above)
Comment (ashoktamang) [inline addon/ng2/utilities/module-resolver.ts]: It was thinking of using this as message after a user uses the command.

File: addon/ng2/utilities/module-resolver.ts (hunk 3 above)
Comment (ashoktamang) [inline addon/ng2/utilities/module-resolver.ts]: Credit goes to Hans and Emmanuel. :)

File: addon/ng2/utilities/module-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        Object.keys(files).forEach(file => {
+          let tempChanges: ReplaceChange[] = files[file]
+            .map(specifier => {
+              let fileDir = path.dirname(file);
+              let changeText = path.relative(fileDir, this.newFilePath);
+              if (changeText.length > 0 && changeText.charAt(0) !== '.') {
+                changeText = `.${path.sep}${changeText}`;
+              }
+              if (changeText === '') {
+                changeText = `.${path.sep}${path.basename(this.oldFilePath, '.ts')}`;
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Why would `changeText` be empty?

File: addon/ng2/utilities/module-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+              } else {
+                changeText = `${changeText}${path.sep}${path.basename(this.oldFilePath, '.ts')}`;
+              }
+              let position = specifier.end - specifier.specifierText.length;
+              return new ReplaceChange(file, position - 1, specifier.specifierText, changeText);
+            });
+          if (tempChanges.length > 1) {
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Don't bother with this check - just call `sort()` regardless.

File: addon/ng2/utilities/module-resolver.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            tempChanges = tempChanges.sort((currentChange, nextChange) => {
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Extract the comparator function you have into a named function - that will make this read a lot better:

`tempChanges.sort(byChangeOrder)`

File: addon/ng2/utilities/module-resolver.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+   * Rewrites the file's own relative imports after it has been moved to new Path.
+   * 
+   * @return {Promise<Change[]>}
+   */
+  resolveOwnImports(): Promise<Change[]> {
+    return dependentFilesUtils.createTsSourceFile(this.oldFilePath)
+      .then((tsFile: ts.SourceFile) => dependentFilesUtils.getImportClauses(tsFile))
+      .then(moduleSpecifiers => {
+        let changes: Change[] = moduleSpecifiers
+        .filter(importClause => {
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Maybe extract this into an `isRelativeImport` method?

File: addon/ng2/utilities/module-resolver.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            return singleSlash || currentDirSyntax || parentDirSyntax;
+          })
+          .map(specifier => {
+            let specifierText = specifier.specifierText;
+            let moduleAbsolutePath = path.resolve(path.dirname(this.oldFilePath), specifierText);
+            let changeText = path.relative(this.newFilePath, moduleAbsolutePath);
+            if (changeText.length > 0 && changeText.charAt(0) !== '.') {
+              changeText = `.${path.sep}${changeText}`;
+            }
+            changeText = (changeText === '') ? `.${path.sep}${path.basename(this.oldFilePath)}` : changeText;
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: No `.ts` extension passed to `path.basename`?

Comment (ashoktamang) [inline addon/ng2/utilities/module-resolver.ts]: I found a way to never get a empty `changeText`, so I have removed this if condition.

File: addon/ng2/utilities/module-resolver.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+              let componentName = path.basename(this.oldFilePath, '.ts');
+              let fileDir = path.dirname(file);
+              let changeText = path.relative(fileDir, path.join(this.newFilePath, componentName));
+              if (changeText.length > 0 && changeText.charAt(0) !== '.') {
+                changeText = `.${path.sep}${changeText}`;
+              };
+              let position = specifier.end - specifier.specifierText.length;
+              return new ReplaceChange(file, position - 1, specifier.specifierText, changeText);
+            });
+            tempChanges.sort((currentChange, nextChange) => nextChange.order - currentChange.order);
```
Comment (alxhub) [inline addon/ng2/utilities/module-resolver.ts]: Do you need this step? Won't they all be sorted before being applied?

Comment (ashoktamang) [inline addon/ng2/utilities/module-resolver.ts]: Yes. removed.

**Comment by hansl** [comment]
LGTM for me. If @alxhub is okay with it then merge it!  Please rebase.

**Comment by sirajc** [comment]
@ashoktamang Can we have some docs on this feature please, or en explanation on how this works

**Pull request merged**